README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

# echor

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/echor)](https://cran.r-project.org/package=echor)
[![echor status
badge](https://mps9506.r-universe.dev/badges/echor)](https://mps9506.r-universe.dev)

[![R build
status](https://github.com/mps9506/echor/workflows/R-CMD-check/badge.svg)](https://github.com/mps9506/echor/actions)
[![Coverage
status](https://codecov.io/gh/mps9506/echor/branch/master/graph/badge.svg)](https://codecov.io/github/mps9506/echor?branch=master)
[![DOI](https://zenodo.org/badge/122131508.svg)](https://zenodo.org/badge/latestdoi/122131508)

<!-- badges: end -->

## Overview

echor downloads wastewater discharge and air emission data for EPA
permitted facilities using the [EPA ECHO API](https://echo.epa.gov/).

## Installation

echor is on CRAN:

    install.packages("echor")

Or install the development version:

    install.packages('echor', repos = 'https://mps9506.r-universe.dev')

## Usage

[Getting
started](https://mps9506.github.io/echor/articles/introduction.html)

[Functions](https://mps9506.github.io/echor/reference/index.html)

## Examples

### Download information about facilities with an NPDES permit

We can look up plants by permit id, bounding box, and numerous other
parameters. I plan on providing documentation of available parameters.
However, arguments can be looked up here:
[get\_cwa\_rest\_services\_get\_facility\_info](https://echo.epa.gov/tools/web-services/facility-search-water#!/Facility_Information/get_cwa_rest_services_get_facility_info)

    library(echor)

    ## echoWaterGetFacilityInfo() will return a dataframe or simple features (sf) dataframe.

    df <- echoWaterGetFacilityInfo(output = "df", 
                                   xmin = '-96.387509', 
                                   ymin = '30.583572', 
                                   xmax = '-96.281422', 
                                   ymax = '30.640008',
                                   p_ptype = "NPD")

    head(df)
    #> # A tibble: 4 × 26
    #>   CWPName Sourc…¹ CWPSt…² CWPCity CWPSt…³ CWPSt…⁴ CWPZip Maste…⁵ Regis…⁶ CWPCo…⁷
    #>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>   <chr>  <chr>   <chr>   <chr>  
    #> 1 CARTER… TX0047… 2200 N… COLLEG… TX      09      77845  <NA>    110070… Brazos 
    #> 2 CENTRA… TX0002… 222 IR… COLLEG… TX      09      77843  <NA>    110009… Brazos 
    #> 3 HEAT T… TX0106… 0.25MI… COLLEG… TX      09      77845  <NA>    110039… Brazos 
    #> 4 TURKEY… TX0062… 3000FT… BRYAN   TX      09      77807  <NA>    110017… Brazos 
    #> # … with 16 more variables: CWPEPARegion <chr>, FacDerivedHuc <chr>,
    #> #   CWPNAICSCodes <chr>, FacLat <dbl>, FacLong <dbl>,
    #> #   CWPTotalDesignFlowNmbr <dbl>, DschToMs4 <chr>, ExposedActivity <chr>,
    #> #   NPDESDataGroupsDescs <chr>, MsgpFacilityInspctnSmmry <chr>,
    #> #   MsgpCorrectiveActionSmmry <chr>, AIRIDs <chr>, NPDESIDs <chr>,
    #> #   SDWAIDs <chr>, CWPDateLastInspSt <date>, BiosolidsFlag <chr>, and
    #> #   abbreviated variable names ¹​SourceID, ²​CWPStreet, ³​CWPState, …

The ECHO database can provide over 270 different columns. echor returns
a subset of these columns that should work for most users. However, you
can specify what data you want returned. Use `echoWaterGetMeta()` to
return a dataframe with column numbers, names, and descriptions to
identify the columns you want returned. Then include the column numbers
as a comma separated string in the `qcolumns` argument. In the example
below, the `qcolumns` argument indicates the dataframe will include
plant name, 8-digit HUC, latitude, longitude, and total design flow.

    df <- echoWaterGetFacilityInfo(output = "df", 
                                   xmin = '-96.387509', 
                                   ymin = '30.583572', 
                                   xmax = '-96.281422', 
                                   ymax = '30.640008',
                                   qcolumns = '1,14,23,24,25',
                                   p_ptype = "NPD")
    head(df)
    #> # A tibble: 4 × 6
    #>   CWPName                SourceID  FacDerivedHuc CWPNAICSCodes FacLat FacLong
    #>   <chr>                  <chr>     <chr>         <chr>          <dbl>   <dbl>
    #> 1 CARTER CREEK WWTP      TX0047163 12070103      <NA>            30.6   -96.3
    #> 2 CENTRAL UTILITY PLANT  TX0002747 12070103      <NA>            30.6   -96.3
    #> 3 HEAT TRANSFER RESEARCH TX0106526 12070101      <NA>            30.6   -96.4
    #> 4 TURKEY CREEK WWTP      TX0062472 12070101      <NA>            30.6   -96.4

When returned as sf dataframes, the data is suitable for immediate
spatial plotting or analysis:

    library(ggspatial)
    library(sf)
    library(ggrepel)
    library(purrr)

    df <- echoWaterGetFacilityInfo(output = "sf", 
                                   xmin = '-96.387509', 
                                   ymin = '30.583572', 
                                   xmax = '-96.281422', 
                                   ymax = '30.640008',
                                   p_ptype = "NPD")

    ##to make labels, need to map the coords and use geom_text :(
    ## can't help but think there is an easier way to do this

    df <- df %>%
      mutate(
        coords = map(geometry, st_coordinates),
        coords_x = map_dbl(coords, 1),
        coords_y = map_dbl(coords, 2)
      )

    ggplot(df) +
      annotation_map_tile(zoomin = -1, progress = "none") +
      geom_sf(inherit.aes = FALSE, shape = 21, 
              color = "darkred", fill = "darkred", 
              size = 2, alpha = 0.25) +
      geom_label_repel(data = df, aes(x = coords_x, y = coords_y, label = SourceID),
                       point.padding = .5, min.segment.length = 0.1,
                       size = 2, color = "dodgerblue") +
      theme_mps_noto() +
      labs(x = "Longitude", y = "Latitude", 
           title = "NPDES permits near Texas A&M",
           caption = "Source: EPA ECHO database")

<img src="man/figures/README-example3-1.png" width="100%" />

### Download discharge/emissions data

Use `echoGetEffluent()` or `echoGetCAAPR()` to download tidy dataframes
of permitted water discharger Discharge Monitoring Report (DMR) or
permitted emitters Clean Air Act annual emissions reports. Please note
that all variables are returned as *character* vectors.

    df <- echoGetEffluent(p_id = 'tx0119407', parameter_code = '00300')

    df <- df %>%
      mutate(dmr_value_nmbr = as.numeric(dmr_value_nmbr),
             monitoring_period_end_date = as.Date(monitoring_period_end_date,
                                                  "%m/%d/%Y")) %>%
      filter(!is.na(dmr_value_nmbr) & limit_value_type_code == "C1")

    ggplot(df) +
      geom_line(aes(monitoring_period_end_date, dmr_value_nmbr)) +
      theme_mps_noto() +
      labs(x = "Monitoring period date",
           y = "Dissolved oxygen concentration (mg/l)",
           title = "Reported minimum dissolved oxygen concentration",
           subtitle = "NPDES ID = TX119407",
           caption = "Source: EPA ECHO")

<img src="man/figures/README-unnamed-chunk-2-1.png" width="672" />

## Session Info

    sessioninfo::platform_info()
    #>  setting  value
    #>  version  R version 4.2.2 (2022-10-31)
    #>  os       Ubuntu 22.04.2 LTS
    #>  system   x86_64, linux-gnu
    #>  ui       X11
    #>  language (EN)
    #>  collate  C.UTF-8
    #>  ctype    C.UTF-8
    #>  tz       UTC
    #>  date     2023-03-12
    #>  pandoc   2.19.2 @ /usr/bin/ (via rmarkdown)
    sessioninfo::package_info()
    #>  ! package      * version    date (UTC) lib source
    #>    abind          1.4-5      2016-07-21 [1] CRAN (R 4.2.2)
    #>    bit            4.0.5      2022-11-15 [1] CRAN (R 4.2.2)
    #>    bit64          4.0.5      2020-08-30 [1] CRAN (R 4.2.2)
    #>  P class          7.3-21     2023-01-23 [?] RSPM (R 4.2.0)
    #>    classInt       0.4-9      2023-02-28 [1] CRAN (R 4.2.2)
    #>    cli            3.6.0      2023-01-09 [1] CRAN (R 4.2.2)
    #>  P codetools      0.2-18     2020-11-04 [?] CRAN (R 4.2.2)
    #>    colorspace     2.1-0      2023-01-23 [1] CRAN (R 4.2.2)
    #>    crayon         1.5.2      2022-09-29 [1] CRAN (R 4.2.2)
    #>    curl           5.0.0      2023-01-12 [1] CRAN (R 4.2.2)
    #>    DBI            1.1.3      2022-06-18 [1] CRAN (R 4.2.2)
    #>    digest         0.6.31     2022-12-11 [1] CRAN (R 4.2.2)
    #>    dplyr        * 1.1.0      2023-01-29 [1] CRAN (R 4.2.2)
    #>    e1071          1.7-13     2023-02-01 [1] CRAN (R 4.2.2)
    #>    echor        * 0.1.7.9000 2023-03-12 [1] local
    #>    ellipsis       0.3.2      2021-04-29 [1] CRAN (R 4.2.2)
    #>    evaluate       0.20       2023-01-17 [1] CRAN (R 4.2.2)
    #>    fansi          1.0.4      2023-01-22 [1] CRAN (R 4.2.2)
    #>    farver         2.1.1      2022-07-06 [1] CRAN (R 4.2.2)
    #>    fastmap        1.1.1      2023-02-24 [1] CRAN (R 4.2.2)
    #>    fs             1.6.1      2023-02-06 [1] CRAN (R 4.2.2)
    #>    generics       0.1.3      2022-07-05 [1] CRAN (R 4.2.2)
    #>    geojsonsf      2.0.3      2022-05-30 [1] CRAN (R 4.2.2)
    #>    ggplot2      * 3.4.1      2023-02-10 [1] CRAN (R 4.2.2)
    #>    ggrepel      * 0.9.3      2023-02-03 [1] CRAN (R 4.2.2)
    #>    ggspatial    * 1.1.7      2022-11-24 [1] CRAN (R 4.2.2)
    #>    glue           1.6.2      2022-02-24 [1] CRAN (R 4.2.2)
    #>    gtable         0.3.1      2022-09-01 [1] CRAN (R 4.2.2)
    #>    highr          0.10       2022-12-22 [1] CRAN (R 4.2.2)
    #>    hms            1.1.2      2022-08-19 [1] CRAN (R 4.2.2)
    #>    htmltools      0.5.4      2022-12-07 [1] CRAN (R 4.2.2)
    #>    httr           1.4.5      2023-02-24 [1] CRAN (R 4.2.2)
    #>    jsonlite       1.8.4      2022-12-06 [1] CRAN (R 4.2.2)
    #>  P KernSmooth     2.23-20    2021-05-03 [?] CRAN (R 4.2.2)
    #>    knitr          1.42       2023-01-25 [1] CRAN (R 4.2.2)
    #>    labeling       0.4.2      2020-10-20 [1] CRAN (R 4.2.2)
    #>  P lattice        0.20-45    2021-09-22 [?] CRAN (R 4.2.2)
    #>    lifecycle      1.0.3      2022-10-07 [1] CRAN (R 4.2.2)
    #>    magrittr       2.0.3      2022-03-30 [1] CRAN (R 4.2.2)
    #>    mpsTemplates * 0.2.0      2023-03-12 [1] Github (mps9506/mpsTemplates@38f84e5)
    #>    munsell        0.5.0      2018-06-12 [1] CRAN (R 4.2.2)
    #>    pillar         1.8.1      2022-08-19 [1] CRAN (R 4.2.2)
    #>    pkgconfig      2.0.3      2019-09-22 [1] CRAN (R 4.2.2)
    #>    plyr           1.8.8      2022-11-11 [1] CRAN (R 4.2.2)
    #>    png            0.1-8      2022-11-29 [1] CRAN (R 4.2.2)
    #>    prettymapr     0.2.4      2022-06-09 [1] CRAN (R 4.2.2)
    #>    prettyunits    1.1.1      2020-01-24 [1] CRAN (R 4.2.2)
    #>    progress       1.2.2      2019-05-16 [1] CRAN (R 4.2.2)
    #>    proxy          0.4-27     2022-06-09 [1] CRAN (R 4.2.2)
    #>    purrr        * 1.0.1      2023-01-10 [1] CRAN (R 4.2.2)
    #>    R6             2.5.1      2021-08-19 [1] CRAN (R 4.2.2)
    #>  P ragg         * 1.2.5      2023-01-12 [?] RSPM (R 4.2.0)
    #>    raster         3.6-20     2023-03-06 [1] CRAN (R 4.2.2)
    #>    Rcpp           1.0.10     2023-01-22 [1] CRAN (R 4.2.2)
    #>    readr          2.1.4      2023-02-10 [1] CRAN (R 4.2.2)
    #>  P renv           0.16.0     2022-09-29 [?] RSPM (R 4.2.0)
    #>    rgdal          1.6-5      2023-03-02 [1] CRAN (R 4.2.2)
    #>    rlang          1.0.6      2022-09-24 [1] CRAN (R 4.2.2)
    #>    rmarkdown      2.20       2023-01-19 [1] CRAN (R 4.2.2)
    #>    rosm           0.2.6      2022-06-09 [1] CRAN (R 4.2.2)
    #>    scales         1.2.1      2022-08-20 [1] CRAN (R 4.2.2)
    #>    sessioninfo    1.2.2      2021-12-06 [1] any (@1.2.2)
    #>    sf           * 1.0-9      2022-11-08 [1] CRAN (R 4.2.2)
    #>    sp             1.6-0      2023-01-19 [1] CRAN (R 4.2.2)
    #>    systemfonts    1.0.4      2022-02-11 [1] CRAN (R 4.2.2)
    #>    terra          1.7-18     2023-03-06 [1] CRAN (R 4.2.2)
    #>  P textshaping    0.3.6      2021-10-13 [?] RSPM (R 4.2.0)
    #>    tibble         3.2.0      2023-03-08 [1] CRAN (R 4.2.2)
    #>    tidyr          1.3.0      2023-01-24 [1] CRAN (R 4.2.2)
    #>    tidyselect     1.2.0      2022-10-10 [1] CRAN (R 4.2.2)
    #>    tzdb           0.3.0      2022-03-28 [1] CRAN (R 4.2.2)
    #>    units          0.8-1      2022-12-10 [1] CRAN (R 4.2.2)
    #>    utf8           1.2.3      2023-01-31 [1] CRAN (R 4.2.2)
    #>    vctrs          0.5.2      2023-01-23 [1] CRAN (R 4.2.2)
    #>    vroom          1.6.1      2023-01-22 [1] CRAN (R 4.2.2)
    #>    withr          2.5.0      2022-03-03 [1] CRAN (R 4.2.2)
    #>    xfun           0.37       2023-01-31 [1] CRAN (R 4.2.2)
    #>    yaml           2.3.7      2023-01-23 [1] CRAN (R 4.2.2)
    #> 
    #>  [1] /home/runner/.cache/R/renv/library/echor-4ec080d0/R-4.2/x86_64-pc-linux-gnu
    #>  [2] /home/runner/work/echor/echor/renv/sandbox/R-4.2/x86_64-pc-linux-gnu/99f2a2ef
    #> 
    #>  P ── Loaded and on-disk path mismatch.


NEWS.md
# echor 0.1.8

## Bug Fixes
* various _getFacilityInfo() functions now properly return records when more then 100,000 records are returned. (fixes #79)
* _getFacilityInfo() functions stop and a message is returned if too many records are requested in sf format.

## Internal changes
* remove use of `rlang::dots_values()` to avoid upcoming soft depreciation. (fixes #77)

# echor 0.1.7

## Minor Changes
* update email address for author and maintainer.
* reduced dependencies by moving spatial examples to pkgdown articles.

## Bug Fixes
* update url endpoint for metadata services.

## Internal changes
* update pkgdown use and template for consistency with my other packages.
* remove rlang::.data selectors for compatibility with new versions of purrr and tidyselect.
* fixed messages introduced by new versions of readr/vroom when columns are NA.
* vignette is no longer built by CRAN to reduce package dependencies, see the pkgdown website for documentation
* remove the depreciated dplyr based progress bar function and use progress package.
* functions provide message and return nothing if nslookup fails.

# echor 0.1.6

## Bug Fixes

* Removed TidyData in DESCRIPTION.
* Fix output for `echoGetCAAPR()`.
* Update base URL used in webservice.

# echor 0.1.5

## Minor Changes

* `httr::RETRY()` used throughout. (fixes #49)
* Add CITATION.
* vignette and readme use ggspatial instead of ggmap. (fixes #52)

# echor 0.1.4

* `downloadDMRs()` passes arguments properly. This fixes #43.


# echor 0.1.3

## Major Changes (possibly breaking)

* `echoGetEffluent()` and `downloadDMRs()` return columns (or nested columns) as characters only.

## Minor changes

* fix vignette (closes #44)
* import tidyr (v1.0.0)
* utilize httptest for unit testing

# echor 0.1.2

## New functions

* `downloadDMRs()` Returns a tidy dataframe with dmr reports. (closes #38)

## Updated functions

* `echoGetEffluent()` uses a different API call to returns a flat dmr report (one plant at a time). Existing arguments remain the same and still returns a dataframe. (closes #37)

# echor 0.1.1

* Single function returns air and water data; `echoGetFacilities()` and `echoGetReports()` wrap the individual functions to streamline functions that users need to be familiar with. (closes #33)
* Added `echoWaterGetMeta()` and `echoAirGetMeta()` functions to retrieve column metadata returned by `echoWaterGetFacilityInfo()` and `echoAirGetFacilityInfo()`; specify columns returned by respective functions by including qcolumns argument in respective functions.   (closes #28).
* Added default `qcolumns` argument to `echoWaterGetFacilityInfo()` to ensure reasonable default columns are returned (#29).
* `echoWaterGetFacilityInfo()` properly returns large datasets. The new internal function `getDownload()` downloads the entire CSV of queried data as a dataframe (closes #27).
* Added default `qcolumns` argument to `echoAirGetFacilityInfo()` to ensure reasonable default columns are returned (#29).
* `echoAirGetFacilityInfo()`  properly returns large datasets. The new internal function `getDownload()` downloads the entire CSV of queried data as a dataframe (closes #27).
* `echoSDWGetSystems()` downloads data for public drinking water systems.
* `echoSDWGetMeta()` downloads metadata for columns returned by `echoSDWGetSystems()`
* Unit tests for functions that rely on EPA ECHO API utilize `skip_on_cran()` instead of `httptest::with_mock_API()`
* Imported functions are called explicitly eg. `pkg::fun()`

# echor 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* Initial Release


tests/testthat.R
library(httptest)
library(echor)

test_check("echor")


tests/testthat/test-echoGetCAAPR.R
context("test-echoGetCAAPR")

library(httptest)
library(here)

.mockPaths(here("tests/testthat/"))

## Not sure if I need to include this in test_that(). Causes test to skip in the specified context.

# capture_requests({
#     echoGetCAAPR(p_id = '110000350174')
#   })


with_mock_api({
  test_that("echoGetCAAPR works", {
    x <- echoGetCAAPR(p_id = '110000350174')
    expect_equal(is.data.frame(x), TRUE)
  })
})


R/echoGetReports.R
#' Downloads self reported discharge and emissions data
#'
#' @param program Character, either \code{program = 'caa'} or \code{program = 'cwa'}. \code{'caa'} retrieves facilities permitted under the Clean Air Act, \code{'cwa'} retrieves facilities permitted under the Clean Water Act.
#' @param p_id Character string specify the identifier for the service. Required.
#' @param verbose Logical, indicating whether to provide processing and retrieval messages. Defaults to FALSE.
#' @param ... Further arguments passed on as query parameters sent to EPA's ECHO API.
#'
#' @return Returns a dataframe
#' @export
echoGetReports <- function(program, p_id, verbose = FALSE, ...) {
  if (program == "caa") {

    echoGetCAAPR(p_id = p_id, verbose = verbose, ...)

  } else if (program == "cwa") {

    echoGetEffluent(p_id = p_id, verbose = verbose, ...)

  } else {

    stop("the argument 'program' must be specified as one of 'caa' or 'cwa' ")

  }
}


R/air.R

# echoAirGetFacilityInfo --------------------------------------------------

#' Downloads EPA ECHO permitted air emitter information
#'
#' Returns a dataframe or simplefeature dataframe of permitted facilities returned by the query.
#' Uses EPA's ECHO API: \url{https://echo.epa.gov/tools/web-services/facility-search-air#!/Facilities/get_air_rest_services_get_facility_info}
#' @param output Character string specifying output format. \code{output = 'df'} for a dataframe or \code{output = 'sf'} for a simple features spatial dataframe. See (\url{https://CRAN.R-project.org/package=sf}) for more information about simple features.
#' @param verbose Logical, indicating whether to provide processing and retrieval messages. Defaults to FALSE
#' @param ... Further arguments passed as query parameters in request sent to EPA ECHO's API. For more options see: \url{https://echo.epa.gov/tools/web-services/facility-search-water#!/Facility_Information/get_air_rest_services_get_facility_info} for a complete list of parameter options. Examples provided below.
#' @import httr
#' @importFrom geojsonsf geojson_sf
#' @return dataframe or sf dataframe suitable for plotting
#' @export
#'
#' @examples\donttest{
#' ## These examples require an internet connection to run
#'
#' ## Retrieve table of facilities by bounding box
#' echoAirGetFacilityInfo(xmin = '-96.407563',
#' ymin = '30.554395',
#' xmax = '-96.25947',
#' ymax = '30.751984',
#' output = 'df')
#'
#' ## Retrieve a simple features dataframe by bounding box
#' spatialdata <- echoAirGetFacilityInfo(xmin = '-96.407563',
#' ymin = '30.554395',
#' xmax = '-96.25947',
#' ymax = '30.751984',
#' output = 'sf')
#'
#' }
echoAirGetFacilityInfo <- function(output = "df", verbose = FALSE, ...) {
    if (length(list(...)) == 0) {
        stop("No valid arguments supplied")
    }
    ## returns a list of arguments supplied by user
    valuesList <- readEchoGetDots(...)

    ## check if user includes an output argument in dots if included, strip it
    ## out
    valuesList <- exclude(valuesList, "output")

    ## check if qcolumns argument is provided by user
    ## if user does not provide qcolumns, provide a sensible default
    if (!("qcolumns" %in% names(valuesList))) {
      qcolumns <- c(1:5,22,23)
      qcolumns <- paste(as.character(qcolumns), collapse = ",")
      valuesList[["qcolumns"]] <- qcolumns
    }

    # check if 1 and 2 are in, if not, insert and order
    valuesList <- insertQColumns(valuesList)

    ## generate query the will be pasted into GET URL
    queryDots <- queryList(valuesList)

    ## build the request URL statement
    path <- "echo/air_rest_services.get_facility_info"
    query <- paste("output=JSON", queryDots, sep = "&")
    getURL <- requestURL(path = path, query = query)

    ## Make the request
    request <- httr::RETRY("GET",
                           url = getURL,
                           httr::accept_json())

    ## Check for valid response for serve, else returns error
    resp_check(request)

    ## Print status message, need to make this optional
    if (isTRUE(verbose)) {
      message("Request URL:", getURL)
      message(httr::http_status(request))
    }

    info <- httr::content(request)

    ## return the query id
    qid <- info[["Results"]][["QueryID"]]

    ## return the number of records
    n_records <- info[["Results"]][["QueryRows"]]
    n_records <- as.numeric(n_records)

    ## build the output

    ## get qcolumns argument specific to this query
    qcolumns <- queryList(valuesList["qcolumns"])

    ## Find out column types so they are parsed correctly
    colNums <- unlist(strsplit(valuesList[["qcolumns"]], split = ","))
    colNums <- as.numeric(colNums)

    colTypes <- columnsToParse(program = "caa", colNums)

    ## if df return output from air_rest_services.get_download
    if (output == "df") {

      if (n_records <= 100000) {

        buildOutput <- getDownload("caa",
                                   qid,
                                   qcolumns,
                                   col_types = colTypes)
      } else {

        # number of pages returned is n_records/5000
        pages <- ceiling(n_records/5000)
        # create the progress bar
        pb <- progress_bar$new(total = pages)

        buildOutput <- getQID("caa",
                              qid,
                              qcolumns,
                              page = 1)
        pb$tick()

        for (i in 2:pages) {
          buildOutput <- bind_rows(buildOutput,
                                   getQID("caa",
                                          qid,
                                          qcolumns,
                                          page = i))
          Sys.sleep(0.5)
          pb$tick()
        }
      }
      return(buildOutput)
    }

    ## if df return output from air_rest_services.get_geojson
    if (output == "sf") {

      ## if returns clusters, there are to many records to
      ## return records via geojson and the request needs to
      ## be more specific. I'm not sure how many records are too
      ## many. If the length of facilities == 0, it means
      ## the query either return no records, or the request returned
      ## clusters and we can stop the function and return a message.
      if(length(info[["Results"]][["Facilities"]]) == 0) {
        if(n_records > 0) {
          stop("Too many records to return spatial a object, please subset your request and try again.")
        }
        if(n_records == 0) {
          stop("No records returned in your request")
        }
      }

      buildOutput <- getGeoJson("caa",
                                qid,
                                qcolumns)
      ## Convert to sf dataframe
      buildOutput <- geojsonsf::geojson_sf(buildOutput)

      return(buildOutput)

      }

    else {
      stop("output argument = ", output,
           ", when it should be either 'df' or 'sf'")
    }

}

# echoAirGetMeta ============================================================

#' Downloads EPA ECHO Air Facility Metadata
#'
#' Returns variable name and descriptions for parameters returned by \code{\link{echoAirGetFacilityInfo}}
#' @param verbose Logical, indicating whether to provide processing and retrieval messages. Defaults to FALSE
#' @import httr
#' @importFrom purrr map_df
#' @return returns a dataframe
#' @export
#'
#' @examples \donttest{
#' ## These examples require an internet connection to run
#'
#' # returns a dataframe of
#' echoAirGetMeta()
#' }
echoAirGetMeta <- function(verbose = FALSE){

  ## build the request URL statement
  path <- "echo/air_rest_services.metadata?output=JSON"
  getURL <- requestURL(path = path, query = NULL)

  ## Make the request
  request <- httr::RETRY("GET",
                         url = getURL,
                         httr::accept_json())

  ## Check for valid response for serve, else returns error
  resp_check(request)

  ## Print status message, need to make this optional
  if (isTRUE(verbose)) {
    message("Request URL:", getURL)
    message(httr::http_status(request))
  }

  info <- httr::content(request)
  info

  ## build the output
  buildOutput <- purrr::map_df(info[["Results"]][["ResultColumns"]],
                               safe_extract,
                               c("ColumnName", "DataType", "DataLength",
                                 "ColumnID", "ObjectName", "Description"))

  return(buildOutput)
}


# echoGetcaapr ------------------------------------------------------------

#' Download EPA ECHO emissions inventory report data
#'
#' @param p_id character string specify the identifier for the service. Required.
#' @param verbose Logical, indicating whether to provide processing and retrieval messages. Defaults to FALSE
#' @param ... Additional arguments
#' @importFrom purrr map_df
#' @importFrom tidyr gather_ pivot_longer
#' @importFrom tibble tibble
#' @import httr
#' @import dplyr
#' @return dataframe
#' @export
#'
#' @examples \donttest{
#' ## This example requires an internet connection to run
#'
#' echoGetCAAPR(p_id = '110000350174')
#' }
#'
echoGetCAAPR <- function(p_id, verbose = FALSE, ...) {

    # check that p_id is a character
    if (!is.character(p_id)) {
        stop("p_id must be a character")
    }


    ## should check if character and return error if not
    p_id <- paste0("p_id=", p_id)

    ## returns a list of arguments supplied by user
    valuesList <- readEchoGetDots(...)

    ## check if user includes an output argument in dots if included, strip it
    ## out
    valuesList <- exclude(valuesList, "output")

    ## generate the intial query
    queryDots <- paste(paste(names(valuesList), valuesList, sep = "="),
                       collapse = "&")

    ## check connectivity
    if (!isTRUE(check_connectivity())) {
      return(invisible(NULL))
    }

    ## build the request URL statement
    path <- "echo/caa_poll_rpt_rest_services.get_caapr"
    query <- paste(p_id, queryDots, sep = "&")
    getURL <- requestURL(path = path, query = query)

    request <- httr::RETRY("GET",
                           url = getURL,
                           httr::accept_json())

    ## Check for valid response for serve, else returns error
    resp_check(request)

    if (isTRUE(verbose)) {
      message("Request URL:", getURL)
      message(httr::http_status(request))
    }

    info <- httr::content(request)

    ## Emissions data is provided in wide format
    pollutant <- purrr::map_df(
      info[["Results"]][["CAAPollRpt"]][["Pollutants"]],
      safe_extract,
      c("Pollutant", "UnitsOfMeasure", "Year1",
        "Year2", "Year3", "Year4",
        "Year5", "Year6", "Year7",
        "Year8", "Year9", "Year10",
        "Program"))
    ## Change emissions data from wide to narrow
#     pollutant <- tidyr::gather_(pollutant, "Year", "Discharge", c("Year1",
#                                                                   "Year2",
#                                                                   "Year3",
#                                                                   "Year4",
#                                                                   "Year5",
#                                                                   "Year6",
#                                                                   "Year7",
#                                                                   "Year8",
#                                                                   "Year9",
#                                                                   "Year10"))
    pollutant <- tidyr::pivot_longer(pollutant,
                                     cols = -c("Pollutant", "UnitsOfMeasure", "Program"),
                                     names_to = "Year",
                                     values_to = "Discharge")


    pollutant <- pollutant %>%
      mutate(Year = case_when(Year == "Year1" ~
                                info[["Results"]][["TRI_year_01"]],
        Year == "Year2" ~ info[["Results"]][["TRI_year_02"]],
        Year == "Year3" ~ info[["Results"]][["TRI_year_03"]],
        Year == "Year4" ~ info[["Results"]][["TRI_year_04"]],
        Year == "Year5" ~ info[["Results"]][["TRI_year_05"]],
        Year == "Year6" ~ info[["Results"]][["TRI_year_06"]],
        Year == "Year7" ~ info[["Results"]][["TRI_year_07"]],
        Year == "Year8" ~ info[["Results"]][["TRI_year_08"]],
        Year == "Year9" ~ info[["Results"]][["TRI_year_09"]],
        Year == "Year10" ~ info[["Results"]][["TRI_year_10"]]))


    ## I should refactor this
    Name <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["FacilityName"]])) {
      NA_character_
    } else {
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["FacilityName"]]
    }
    SourceID <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["SourceId"]])) {
      NA_character_
    } else {
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["SourceId"]]
    }
    Street <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Street"]])) {
      NA_character_
    } else {
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Street"]]
    }
    City <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["City"]])) {
      NA_character_
    } else {
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["City"]]
    }
    State <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["State"]])) {
      NA_character_
    } else{
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["State"]]
    }
    Zip <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Zip"]])) {
      NA_character_
    } else {
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Zip"]]
    }
    County <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["County"]])) {
      NA_character_
    } else {
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["County"]]
    }
    Region <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Region"]])) {
      NA_character_
    } else {
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Region"]]
    }
    Latitude <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Latitude"]])) {
      NA
    } else {
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Latitude"]]
    }
    Longitude <- if(is.null(info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Longitude"]])) {
      NA
    } else {
      info[["Results"]][["CAAPollRpt"]][["RegistryIDs"]][[1]][["Longitude"]]
    }

    ## build output dataframe
    buildOutput <- tibble::tibble(
      Name = Name,
      SourceID = SourceID,
      Street = Street,
      City = City,
      State = State,
      Zip = Zip,
      County = County,
      Region = Region,
      Latitude = as.numeric(Latitude),
      Longitude = as.numeric(Longitude),
      Pollutant = as.factor(pollutant$Pollutant),
      UnitsOfMeasure = as.factor(pollutant$UnitsOfMeasure),
      Program = as.factor(pollutant$Program),
      Year = as.integer(pollutant$Year),
      Discharge = as.numeric(gsub(",", "", pollutant$Discharge))
    )

    return(buildOutput)

}
